### PR TITLE
Pickle in Union Type

### PR DESCRIPTION
**Pull request by pingsutw**
# TL;DR
Problems:
- Sequence wasn't converted to pickle
- Unknown type in Union wasn't converted to pickle

Both `Sequence` and `Union` have `__origin__` attribute, but we didn't handle these cases in [_change_unrecognized_type_to_pickle](https://github.com/flyteorg/flytekit/blob/22580d8ba63fe1842a9eb0de6281b2d5eb4bd588/flytekit/core/interface.py#L261-L265) 

## Type
 - [x] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
https://github.com/flyteorg/flyte/issues/2823

## Follow-up issue
_NA_

### CODE REVIEW

File: flytekit/core/interface.py [hunk 1]
```diff
@@ -329,7 +333,10 @@ def transform_variable_map(
                 elif v.__origin__ is dict:
                     sub_type = v.__args__[1]
             if hasattr(sub_type, "__origin__") and sub_type.__origin__ is FlytePickle:
-                res[k].type.metadata = {"python_class_name": sub_type.python_type().__name__}
+                if hasattr(sub_type.python_type(), "__name__"):
+                    res[k].type.metadata = {"python_class_name": sub_type.python_type().__name__}
+                elif hasattr(sub_type.python_type(), "_name"):
```
Comment (wild-endeavor) [inline flytekit/core/interface.py:338]: this is cryptic, can you add a comment here please?

File: flytekit/core/interface.py [hunk 2]
```diff
@@ -259,12 +259,16 @@ def transform_interface_to_list_interface(interface: Interface) -> Interface:
...
         if hasattr(t, "__origin__") and hasattr(t, "__args__"):
-            if t.__origin__ == list:
+            if get_origin(t) == list:
                 return typing.List[_change_unrecognized_type_to_pickle(t.__args__[0])]
-            elif t.__origin__ == dict and t.__args__[0] == str:
+            elif get_origin(t) == dict and t.__args__[0] == str:
                 return typing.Dict[str, _change_unrecognized_type_to_pickle(t.__args__[1])]
-        else:
-            TypeEngine.get_transformer(t)
+            elif get_origin(t) == typing.Union:
```
Comment (wild-endeavor) [inline flytekit/core/interface.py:266]: can we use `is` instead of == or does that not work here?

**Review by wild-endeavor — DISMISSED** [review]

File: flytekit/core/interface.py (hunk 1 above)
Comment (pingsutw) [reply flytekit/core/interface.py:338]: sorry, added it.

File: flytekit/core/interface.py (hunk 2 above)
Comment (pingsutw) [reply flytekit/core/interface.py]: updated it.

**Review by wild-endeavor — APPROVED** [review]
